### PR TITLE
User language in cookies

### PR DESCRIPTION
**Pull request by iborsch**
Now we can save user choice of interface language to cookies.
We define user language as the chosen language in case it exists. Otherwise, we set user language to the default one.

There are several options:
1. If a request contains locale:

   1. If there are no cookies or locale is the same as the user language, we just handle the request.

   2. If locale differs from the user language, we replace locale with the user language.

2. If a request doesn't contain locale, we add user language to the request.

### CODE REVIEW

**Comment by iborsch** [comment]
[save selected lang to cookie](https://trello.com/c/Hwx1qwNc/321-save-selected-lang-to-cookie)

File: server.js [hunk 1]
```diff
@@ -18,6 +18,9 @@ const app = next({ dev });
 
 const handle = routes.getRequestHandler(app);
 
+const hasLangCookie = req => req.cookies && req.cookies[LANG_COOKIE_NAME];
```
Comment (Drapegnik) [inline server.js]: is it possible that `req.cookies` doesn't exist?

File: server.js [hunk 2]
```diff
@@ -18,6 +18,9 @@ const app = next({ dev });
...
+const getUserLang = req => (hasLangCookie(req) ? req.cookies[LANG_COOKIE_NAME] : DEFAULT_LOCALE);
```
Comment (Drapegnik) [inline server.js]:
```suggestion
const getUserLang = req => req.cookies[LANG_COOKIE_NAME] || DEFAULT_LOCALE;
```

File: server.js [hunk 3]
```diff
@@ -37,14 +41,26 @@ app.prepare().then(() => {
...
+    const userLang = getUserLang(req);
+
+    if (STATIC_PATHS.includes(startPath)) {
       return handle(req, res);
     }
-    // missed locale, add `be` & redirect
-    return res.redirect(`/${DEFAULT_LOCALE}${req.originalUrl}`);
+
+    if (VALID_LOCALES.includes(startPath)) {
+      if (userLang === startPath || !hasLangCookie(req)) {
```
Comment (Drapegnik) [inline server.js]: `userLang === startPath` seems that this is always `false`

Comment (Drapegnik) [reply server.js]: maybe `startPath.startsWith(userLang)`

File: server.js [hunk 4]
```diff
@@ -37,14 +41,26 @@ app.prepare().then(() => {
...
+        return handle(req, res);
+      }
+
+      return res.redirect(`/${userLang}${req.params['0']}`);
```
Comment (Drapegnik) [inline server.js]: what is `req.params['0']`?

**Review by Drapegnik — CHANGES_REQUESTED** [review]

File: server.js (hunk 1 above)
Comment (iborsch) [reply server.js]: hm, seems it is not.

File: server.js (hunk 4 above)
Comment (iborsch) [reply server.js]: It is a part of the url that follows startPath. In the case above, req.params['0'] would be equal to '/about'.

File: server.js (hunk 3 above)
Comment (iborsch) [reply server.js]: startPath is the first parameter of the url. For example, in the case of wir.by/en/about, startPath would be equal to 'en'.

File: pages/_app.js [hunk 5]
```diff
@@ -68,7 +68,9 @@ class Root extends App {
     const { Component, pageProps, store, router, user } = this.props;
     const getLayoutProps = Component.getLayoutProps || getEmptyObject;
     const { lang } = router.query;
-    const locale = lang || DEFAULT_LOCALE;
+    // asPath starts with /, so we have to take [1], not [0]
+    const parsedLang = router.asPath.split('/')[1];
+    const locale = lang || (VALID_LOCALES.includes(parsedLang) && parsedLang) || DEFAULT_LOCALE;
```
Comment (Drapegnik) [inline pages/_app.js]: lets extract this to function:
```js
const getLocale = ({ asPath, query: { lang } }) => {
	if (lang) {
		return lang;
	}
	const parsedLang = asPath.split('/')[1];
	// please add comment about this case
	if (VALID_LOCALES.includes(parsedLang)) {
		return parsedLang
	}
	return DEFAULT_LOCALE;
}

const locale = getLocale(router)

File: routes.js [hunk 6]
```diff
@@ -1,9 +1,9 @@
 const { cloneElement } = require('react');
 const routes = require('next-routes')();
 
-const { LOCALES } = require('./constants');
+const { VALID_LOCALES } = require('./constants');
 
-const langs = Object.keys(LOCALES).join('|');
+const langs = VALID_LOCALES.join('|');
```
Comment (Drapegnik) [inline routes.js:6]: 👍

File: server.js [hunk 7]
```diff
@@ -37,14 +38,26 @@ app.prepare().then(() => {
   // This is useful for fully local development.
   server.use('/test', proxy({ target: BACKEND_URL, changeOrigin: true }));
 
-  server.get('/', (req, res) => res.redirect(`/${DEFAULT_LOCALE}/`));
+  server.get('/', (req, res) => res.redirect(`/${getUserLang(req)}/`));
+
   server.get('/:startPath*', (req, res) => {
     const { startPath } = req.params;
-    if (VALID_PATHS.includes(startPath)) {
+    const userLang = getUserLang(req);
```
Comment (Drapegnik) [inline server.js]: what if `req.cookies[LANG_COOKIE_NAME] === 'wrong_lang'`?

File: server.js [hunk 8]
```diff
@@ -37,14 +38,26 @@ app.prepare().then(() => {
...
+    if (VALID_LOCALES.includes(startPath)) {
+      if (userLang === startPath || !req.cookies[LANG_COOKIE_NAME]) {
+        return handle(req, res);
+      }
+
+      return res.redirect(`/${userLang}${req.params[0]}`);
+    }
+
+    // missed locale, add user language & redirect
+    return res.redirect(`/${userLang}${req.originalUrl}`);
```
Comment (Drapegnik) [inline server.js]: you can lift up this case and make branching flat (without nested `if`s)

Comment (Drapegnik) [reply server.js]:
```js
if (!VALID_LOCALES.includes(startPath)) {
    // missed locale, add user language & redirect
    return res.redirect(`/${userLang}${req.originalUrl}`);
}
```

**Review by Drapegnik — CHANGES_REQUESTED** [review]

**Review by Drapegnik — APPROVED** [review]

**Pull request merged**